### ISSUE TITLE
Invalid clang arguments generated in ninja backend for cmake C subproject on Windows 

### ISSUE THREAD

**Issue by rminderhoud** [mesonbuild/meson#9225]
**Describe the bug**
Windows 10 (x64)
clang version 10.0.0
meson 0.59.1

I am using GLFW 3.3.4 (C project) in my C++ project by using the CMake module and keeping GLFW as a subproject in my source tree. Meson correctly detects and configures this target but compilation fails because there is an argument in the argument list that causes clang to error out with this error:

`clang: error: no such file or directory: '/clang:-Wall'`

<details><summary>Full Error</summary>
<pre><code>
PS Z:\gdb-viewer\build> ninja
[0/1] Regenerating build files.
The Meson build system
Version: 0.59.1
Source dir: Z:\gdb-viewer
Build dir: Z:\gdb-viewer\build
Build type: native build
Project name: Rednim GDB Viewer
Project version: 0.1.0
C++ compiler for the host machine: clang++ (clang 10.0.0 "clang version 10.0.0 ")
C++ linker for the host machine: clang++ link 14.29.30133.0
Host machine cpu family: x86_64
Host machine cpu: x86_64

Executing subproject glfw-3.3.4 method cmake

glfw-3.3.4| Found CMake: C:\Users\Ralph\scoop\shims\cmake.EXE (3.21.2)

| Configuring the build directory with CMake version 3.21.2
| Running CMake with: -G Ninja -DCMAKE_INSTALL_PREFIX=c:\
|   - build directory:          Z:/gdb-viewer/build/subprojects/glfw-3.3.4/__CMake_build
|   - source directory:         Z:/gdb-viewer/subprojects/glfw-3.3.4
|   - toolchain file:           Z:/gdb-viewer/build/subprojects/glfw-3.3.4/__CMake_build/CMakeMesonToolchainFile.cmake
|   - preload file:             Z:/gdb-viewer/build/meson-private/data/preload.cmake
|   - trace args:               --trace-expand --trace-format=json-v1 --no-warn-unused-cli --trace-redirect=cmake_trace.txt
|   - disabled policy warnings: [CMP0025, CMP0047, CMP0056, CMP0060, CMP0065, CMP0066, CMP0067, CMP0082, CMP0089, CMP0102]

| Running with expanded trace output on.
| Not searching for unused variables given on the command line.
| Trace will be written to cmake_trace.txt
| -- The C compiler identification is Clang 10.0.0 with GNU-like command-line
| -- Detecting C compiler ABI info
| -- Detecting C compiler ABI info - done
| -- Check for working C compiler: C:/Program Files/LLVM/bin/clang.EXE - skipped
| -- Detecting C compile features
| -- Detecting C compile features - done
| -- Looking for pthread.h
| -- Looking for pthread.h - not found
| -- Found Threads: TRUE
| -- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE)
| -- Using Win32 for window creation
| -- Configuring done
| -- Generating done
| -- Build files have been written to: Z:/gdb-viewer/build/subprojects/glfw-3.3.4/__CMake_build

glfw-3.3.4| CMake configuration: SUCCEEDED
glfw-3.3.4| CMake project GLFW has 31 build targets.

cmake-ast| Processing generated meson AST
cmake-ast| Build file: Z:\gdb-viewer\build\subprojects\glfw-3.3.4\meson.build

glfw-3.3.4| Project name: GLFW
glfw-3.3.4| Project version: undefined
glfw-3.3.4| C compiler for the host machine: clang (clang 10.0.0 "clang version 10.0.0 ")
glfw-3.3.4| C linker for the host machine: clang link 14.29.30133.0
glfw-3.3.4| Program C:\Users\Ralph\AppData\Roaming\Python\Python39\Scripts\meson found: YES (C:\Users\Ralph\AppData\Roaming\Python\Python39\Scripts\meson.exe)
glfw-3.3.4| Build targets in project: 31
glfw-3.3.4| Subproject glfw-3.3.4 finished.

Build targets in project: 32

Rednim GDB Viewer 0.1.0

  Subprojects
    glfw-3.3.4: YES

Found ninja-1.10.2 at C:\Users\Ralph\scoop\shims\ninja.EXE
[1/18] Compiling C object subprojects/glfw-3.3.4/libglfw.a.p/src_wgl_context.c.obj
FAILED: subprojects/glfw-3.3.4/libglfw.a.p/src_wgl_context.c.obj
"clang" "-Isubprojects\glfw-3.3.4\libglfw.a.p" "-Isubprojects\glfw-3.3.4" "-I..\subprojects\glfw-3.3.4" "-I..\subprojects\glfw-3.3.4\include" "-I..\subprojects\glfw-3.3.4\src" "-Isubprojects\glfw-3.3.4\__CMake_build\src" "-I..\subprojects\glfw-3.3.4\__CMake_build\src" "-Isubprojects\glfw-3.3.4\__CMake_build" "-I..\subprojects\glfw-3.3.4\__CMake_build" "-fcolor-diagnostics" "-D_FILE_OFFSET_BITS=64" "-Wall" "-Winvalid-pch" "-std=gnu99" "-O0" "-g" "-g" "-Xclang" "-gcodeview" "-O0" "-D_DEBUG" "-D_DLL" "-D_MT" "-Xclang" "--dependent-lib=msvcrtd" "/clang:-Wall" "-D_GLFW_USE_CONFIG_H" "-D_UNICODE" -MD -MQ subprojects/glfw-3.3.4/libglfw.a.p/src_wgl_context.c.obj -MF "subprojects\glfw-3.3.4\libglfw.a.p\src_wgl_context.c.obj.d" -o subprojects/glfw-3.3.4/libglfw.a.p/src_wgl_context.c.obj "-c" ../subprojects/glfw-3.3.4/src/wgl_context.c
clang: error: no such file or directory: '/clang:-Wall'
[2/18] Compiling C object subprojects/glfw-3.3.4/libglfw.a.p/src_win32_monitor.c.obj
FAILED: subprojects/glfw-3.3.4/libglfw.a.p/src_win32_monitor.c.obj
"clang" "-Isubprojects\glfw-3.3.4\libglfw.a.p" "-Isubprojects\glfw-3.3.4" "-I..\subprojects\glfw-3.3.4" "-I..\subprojects\glfw-3.3.4\include" "-I..\subprojects\glfw-3.3.4\src" "-Isubprojects\glfw-3.3.4\__CMake_build\src" "-I..\subprojects\glfw-3.3.4\__CMake_build\src" "-Isubprojects\glfw-3.3.4\__CMake_build" "-I..\subprojects\glfw-3.3.4\__CMake_build" "-fcolor-diagnostics" "-D_FILE_OFFSET_BITS=64" "-Wall" "-Winvalid-pch" "-std=gnu99" "-O0" "-g" "-g" "-Xclang" "-gcodeview" "-O0" "-D_DEBUG" "-D_DLL" "-D_MT" "-Xclang" "--dependent-lib=msvcrtd" "/clang:-Wall" "-D_GLFW_USE_CONFIG_H" "-D_UNICODE" -MD -MQ subprojects/glfw-3.3.4/libglfw.a.p/src_win32_monitor.c.obj -MF "subprojects\glfw-3.3.4\libglfw.a.p\src_win32_monitor.c.obj.d" -o subprojects/glfw-3.3.4/libglfw.a.p/src_win32_monitor.c.obj "-c" ../subprojects/glfw-3.3.4/src/win32_monitor.c
clang: error: no such file or directory: '/clang:-Wall'
[3/18] Compiling C object subprojects/glfw-3.3.4/libglfw.a.p/src_osmesa_context.c.obj
FAILED: subprojects/glfw-3.3.4/libglfw.a.p/src_osmesa_context.c.obj
"clang" "-Isubprojects\glfw-3.3.4\libglfw.a.p" "-Isubprojects\glfw-3.3.4" "-I..\subprojects\glfw-3.3.4" "-I..\subprojects\glfw-3.3.4\include" "-I..\subprojects\glfw-3.3.4\src" "-Isubprojects\glfw-3.3.4\__CMake_build\src" "-I..\subprojects\glfw-3.3.4\__CMake_build\src" "-Isubprojects\glfw-3.3.4\__CMake_build" "-I..\subprojects\glfw-3.3.4\__CMake_build" "-fcolor-diagnostics" "-D_FILE_OFFSET_BITS=64" "-Wall" "-Winvalid-pch" "-std=gnu99" "-O0" "-g" "-g" "-Xclang" "-gcodeview" "-O0" "-D_DEBUG" "-D_DLL" "-D_MT" "-Xclang" "--dependent-lib=msvcrtd" "/clang:-Wall" "-D_GLFW_USE_CONFIG_H" "-D_UNICODE" -MD -MQ subprojects/glfw-3.3.4/libglfw.a.p/src_osmesa_context.c.obj -MF "subprojects\glfw-3.3.4\libglfw.a.p\src_osmesa_context.c.obj.d" -o subprojects/glfw-3.3.4/libglfw.a.p/src_osmesa_context.c.obj "-c" ../subprojects/glfw-3.3.4/src/osmesa_context.c
clang: error: no such file or directory: '/clang:-Wall'
[4/18] Compiling C object subprojects/glfw-3.3.4/libglfw.a.p/src_win32_time.c.obj
FAILED: subprojects/glfw-3.3.4/libglfw.a.p/src_win32_time.c.obj
"clang" "-Isubprojects\glfw-3.3.4\libglfw.a.p" "-Isubprojects\glfw-3.3.4" "-I..\subprojects\glfw-3.3.4" "-I..\subprojects\glfw-3.3.4\include" "-I..\subprojects\glfw-3.3.4\src" "-Isubprojects\glfw-3.3.4\__CMake_build\src" "-I..\subprojects\glfw-3.3.4\__CMake_build\src" "-Isubprojects\glfw-3.3.4\__CMake_build" "-I..\subprojects\glfw-3.3.4\__CMake_build" "-fcolor-diagnostics" "-D_FILE_OFFSET_BITS=64" "-Wall" "-Winvalid-pch" "-std=gnu99" "-O0" "-g" "-g" "-Xclang" "-gcodeview" "-O0" "-D_DEBUG" "-D_DLL" "-D_MT" "-Xclang" "--dependent-lib=msvcrtd" "/clang:-Wall" "-D_GLFW_USE_CONFIG_H" "-D_UNICODE" -MD -MQ subprojects/glfw-3.3.4/libglfw.a.p/src_win32_time.c.obj -MF "subprojects\glfw-3.3.4\libglfw.a.p\src_win32_time.c.obj.d" -o subprojects/glfw-3.3.4/libglfw.a.p/src_win32_time.c.obj "-c" ../subprojects/glfw-3.3.4/src/win32_time.c
clang: error: no such file or directory: '/clang:-Wall'
[5/18] Compiling C object subprojects/glfw-3.3.4/libglfw.a.p/src_win32_window.c.obj
FAILED: subprojects/glfw-3.3.4/libglfw.a.p/src_win32_window.c.obj
"clang" "-Isubprojects\glfw-3.3.4\libglfw.a.p" "-Isubprojects\glfw-3.3.4" "-I..\subprojects\glfw-3.3.4" "-I..\subprojects\glfw-3.3.4\include" "-I..\subprojects\glfw-3.3.4\src" "-Isubprojects\glfw-3.3.4\__CMake_build\src" "-I..\subprojects\glfw-3.3.4\__CMake_build\src" "-Isubprojects\glfw-3.3.4\__CMake_build" "-I..\subprojects\glfw-3.3.4\__CMake_build" "-fcolor-diagnostics" "-D_FILE_OFFSET_BITS=64" "-Wall" "-Winvalid-pch" "-std=gnu99" "-O0" "-g" "-g" "-Xclang" "-gcodeview" "-O0" "-D_DEBUG" "-D_DLL" "-D_MT" "-Xclang" "--dependent-lib=msvcrtd" "/clang:-Wall" "-D_GLFW_USE_CONFIG_H" "-D_UNICODE" -MD -MQ subprojects/glfw-3.3.4/libglfw.a.p/src_win32_window.c.obj -MF "subprojects\glfw-3.3.4\libglfw.a.p\src_win32_window.c.obj.d" -o subprojects/glfw-3.3.4/libglfw.a.p/src_win32_window.c.obj "-c" ../subprojects/glfw-3.3.4/src/win32_window.c
clang: error: no such file or directory: '/clang:-Wall'
[6/18] Compiling C object subprojects/glfw-3.3.4/libglfw.a.p/src_win32_thread.c.obj
FAILED: subprojects/glfw-3.3.4/libglfw.a.p/src_win32_thread.c.obj
"clang" "-Isubprojects\glfw-3.3.4\libglfw.a.p" "-Isubprojects\glfw-3.3.4" "-I..\subprojects\glfw-3.3.4" "-I..\subprojects\glfw-3.3.4\include" "-I..\subprojects\glfw-3.3.4\src" "-Isubprojects\glfw-3.3.4\__CMake_build\src" "-I..\subprojects\glfw-3.3.4\__CMake_build\src" "-Isubprojects\glfw-3.3.4\__CMake_build" "-I..\subprojects\glfw-3.3.4\__CMake_build" "-fcolor-diagnostics" "-D_FILE_OFFSET_BITS=64" "-Wall" "-Winvalid-pch" "-std=gnu99" "-O0" "-g" "-g" "-Xclang" "-gcodeview" "-O0" "-D_DEBUG" "-D_DLL" "-D_MT" "-Xclang" "--dependent-lib=msvcrtd" "/clang:-Wall" "-D_GLFW_USE_CONFIG_H" "-D_UNICODE" -MD -MQ subprojects/glfw-3.3.4/libglfw.a.p/src_win32_thread.c.obj -MF "subprojects\glfw-3.3.4\libglfw.a.p\src_win32_thread.c.obj.d" -o subprojects/glfw-3.3.4/libglfw.a.p/src_win32_thread.c.obj "-c" ../subprojects/glfw-3.3.4/src/win32_thread.c
clang: error: no such file or directory: '/clang:-Wall'
ninja: build stopped: subcommand failed.
</code></pre>
</details>

**To Reproduce**
```
project(
    'Rednim GDB Viewer',
    'cpp',
    version: '0.1.0'
)

# Import the cmake module and make available in our project
cmake = import('cmake')

glfw_proj = cmake.subproject('glfw-3.3.4')
glfw_dep = glfw_proj.dependency('glfw')

executable('gdb-viewer', 'src/main.cpp', dependencies: [glfw_dep])
```

**Expected behavior**
The expected behavior is this should work out of the box. Removing all instances of `"/clang:-Wall"` in the `build.ninja` file fixes the issue and the subproject can be compiled.

**system parameters**
Windows 10 (x64)
clang version 10.0.0
meson 0.59.1

**Comment by dcbaker** [comment]
Are you using the clang-cl driver, or the gcc-esqe clang? This might be a bug in the way the clang abstraction works on windows

**Comment by mensinda** [comment]
If this is a CMake issue, we need the `meson-log.txt`. Additionally from `build/subprojects/<your SP>` we need the `meson.build`, `fileAPI.json`, and the `__CMake_build` directory.

**Comment by mensinda** [comment]
Additionally, the project in `build/subprojects/<your SP>/__CMake_build` should be fully buildable by ninja (or msbuild depending on your setup). If it isn't, there might be a CMake / CMakeLists.txt issue.

**Comment by rminderhoud** [comment]
@dcbaker I'm using clang directly (not GCC) but I'm not sure if that is what you are asking.

@mensinda Running `ninja` from that `__CMake_build` returns the same error. I've attached the files you've requested (although I renamed the extensions so github wouldn't complain).

I appreciate you both taking a look at this issue. Please let me know how else I can help troubleshoot, unfortunately I am not too familiar with meson internals at this time to provide better information.

[meson-log.txt](https://github.com/mesonbuild/meson/files/7117186/meson-log.txt)
[meson.build.txt](https://github.com/mesonbuild/meson/files/7117189/meson.build.txt)
[fileAPI.json.txt](https://github.com/mesonbuild/meson/files/7117191/fileAPI.json.txt)

**Comment by mensinda** [comment]
@rminderhoud I still need the full `__CMake_build` build directory just to be sure. However, from line 164 in the `meson-log.txt` it is clear that the `/clang:-Wall` is already coming from CMake, which also explains why running `ninja` in `__CMake_build` fails. Note, that the `build.ninja` was generated from CMake and not Meson.

Could you try building GLFW3 with CMake and clang directly? If the same error occurs again then this is GLFW3 / CMake problem and unrelated to Meson.

**Comment by rminderhoud** [comment]
@mensinda You are correct. This is a GLFW3/ CMake issue as when I try to build the project directly using CMake and then running Ninja it errors out in the exact same way. Thanks for talking a look, sorry for the false alarm.

**Comment by rminderhoud** [comment]
For anyone who happens across this issue for some reason, the fix went in here https://github.com/glfw/glfw/commit/061a0263a9783c1442ad96a061c717c167ab4a76 and is only available in GLFW 3.3.5 (not yet released)
